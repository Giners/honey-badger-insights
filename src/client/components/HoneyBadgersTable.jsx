import React from 'react'
import PropTypes from 'prop-types'

import { PagingState, IntegratedPaging } from '@devexpress/dx-react-grid'

import {
  Grid,
  Table,
  TableHeaderRow,
  PagingPanel,
} from '@devexpress/dx-react-grid-material-ui'

/**
 * Takes a list of honey badgers and their details and turns it into rows of details for use in the
 * table that displays the details about honey badgers. It presumes that at a minimum that the
 * IP address ('ipAddress') and the amount of times the honey badger has been seen in the last 24
 * hours ('count') are present on the details.
 */
const getHoneyBadgersRows = honeyBadgers =>
  honeyBadgers.map(honeyBadger => {
    // Supply the minimum amount of info we expect in a row which is the IP address and the amount
    // of times we have seen the entity in the last 24 hours. Add a 'Loading' flag as the default
    // for all of the other info that may not be loaded yet.
    const { ipAddress, count } = honeyBadger
    const row = { ipAddress, count, asn: 'Loading' }

    // Add the autonomous system number if autonomous system info has been loaded...
    if (honeyBadger.as) {
      row.asn = honeyBadger.as.asn
    }

    return row
  })

const HoneyBadgersTable = ({ honeyBadgers }) => (
  <Grid
    rows={getHoneyBadgersRows(honeyBadgers)}
    columns={[
      { name: 'ipAddress', title: 'IP Address' },
      { name: 'count', title: 'Times Seen Last 24 Hours' },
      { name: 'asn', title: 'Autonomous System #' },
    ]}
  >
    <PagingState defaultCurrentPage={0} pageSize={5} />
    <IntegratedPaging />
    <Table />
    <TableHeaderRow />
    <PagingPanel />
  </Grid>
)

HoneyBadgersTable.propTypes = {
  honeyBadgers: PropTypes.arrayOf(
    PropTypes.shape({
      ipAddress: PropTypes.string,
      count: PropTypes.number,
    }),
  ).isRequired,
}

export default HoneyBadgersTable
