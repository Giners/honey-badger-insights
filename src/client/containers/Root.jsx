import React from 'react'
import ReactDOM from 'react-dom'

const Root = () => 'LOL Bananas are silly'

// Use 'ReactDOM.render()' instead of 'ReactDOM.hydrate()' as we aren't doing any SSR/weren't
// rendered with 'ReactDOMServer.renderToString()'
ReactDOM.render(<Root />, document.getElementById('root'))
