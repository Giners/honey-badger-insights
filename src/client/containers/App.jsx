import 'isomorphic-fetch'
import { ApolloClient } from 'apollo-client'
import { HttpLink } from 'apollo-link-http'
import { InMemoryCache } from 'apollo-cache-inmemory'

import React from 'react'

import HoneyBadgersTable from './../components/HoneyBadgersTable'
import {
  autonomousSystemsQuery,
  geoLocationsQuery,
  topHoneyBadgersQuery,
} from './../graphql/queries'
import config from './../../server/config'

const apolloClient = new ApolloClient({
  link: new HttpLink({
    uri: `${config.app.hostURL}:${config.app.port}${
      config.serviceURIs.graphQL
    }`,
  }),
  cache: new InMemoryCache(),
})

/** The initial state of the App component when it is mounted. Exported for testing purposes. */
export const initialState = { honeyBadgers: new Map(), error: null }

/**
 * The <App> component is a stateful React component that is responsible for providing the "root"
 * for our visual and logical React components that make up our app. Its most important job is to
 * query for information about honey badgers and then to pass that information along to child
 * components so they can render it into a meaningful visualization.
 */
class App extends React.Component {
  /**
   * Helper method that makes a query to the GraphQL service/API endpoint for the top honey badgers
   * that have been seen in the past 24 hours.
   *
   * @return {Promise<Map<string, HoneyBadger>, error>} Returns a promise where it is resolved to a
   * map that contains info about the top honey badgers. The map is keyed off of the honey badgers
   * IP address. Should this fail to query the GraphQL service/API endpoint or for any other reason
   * this promise will be rejected with the error.
   */
  static async collectTopHoneyBadgersInfo() {
    const { topHoneyBadgers } = (await apolloClient.query({
      query: topHoneyBadgersQuery,
    })).data

    const thbKVPs = topHoneyBadgers.map(topHoneyBadger => [
      topHoneyBadger.ipAddress,
      topHoneyBadger,
    ])
    return new Map(thbKVPs)
  }

  /**
   * Helper method that given a map of honey badgers where they are keyed off of their IP address
   * will make a query to the GraphQL service/API endpoint for said honey badger for info about the
   * autonomous systems that is responsible/in charge of said honey badger.
   *
   * Note that this function isn't pure and has side effects in that it modifies the map of honey
   * badgers passed into it to add in the info about the autonomous systems that has been
   * collected.
   *
   * @param {Map<string, HoneyBadger>} honeyBadgers - Map of honey badgers where they are keyed off
   * of their IP address
   *
   * @return {Promise<undefined, error>} Returns a promise where it will resolve if able to
   * successfully query for info about autonomous systems as well as add the info to the map of
   * honey badgers that is passed into it. Should we fail to query the GraphQL service/API endpoint
   * or for any other reason this promise will be rejected with the error.
   */
  static async collectAutonomousSystemsInfo(honeyBadgers) {
    const ipAddresses = [...honeyBadgers.keys()]

    const { autonomousSystems } = (await apolloClient.query({
      query: autonomousSystemsQuery,
      variables: { ipAddresses },
    })).data

    // Join the data about the autonomous system that the honey badger falls under with the
    // existing data about honey badgers
    autonomousSystems.forEach(({ ipAddress, ...rest }) => {
      const honeyBadger = honeyBadgers.get(ipAddress)

      honeyBadgers.set(ipAddress, { ...honeyBadger, as: { ...rest } })
    })
  }

  /**
   * Helper method that given a map of honey badgers where they are keyed off of their IP address
   * will make a query to the GraphQL service/API endpoint for said honey badger for info about
   * where said honey badger is located geospatially.
   *
   * Note that this function isn't pure and has side effects in that it modifies the map of honey
   * badgers passed into it to add in the info about the locations that has been collected.
   *
   * @param {Map<string, HoneyBadger>} honeyBadgers - Map of honey badgers where they are keyed off
   * of their IP address
   *
   * @return {Promise<undefined, error>} Returns a promise where it will resolve if able to
   * successfully query for info about the locations as well as add the info to the map of honey
   * badgers that is passed into it. Should we fail to query the GraphQL service/API endpoint
   * or for any other reason this promise will be rejected with the error.
   */
  static async collectGeoLocationsInfo(honeyBadgers) {
    const ipAddresses = [...honeyBadgers.keys()]

    const { geoLocations } = (await apolloClient.query({
      query: geoLocationsQuery,
      variables: { ipAddresses },
    })).data

    // Join the data about the location of the honey badger with the existing data about honey
    // badgers
    geoLocations.forEach(({ ipAddress, ...rest }) => {
      const honeyBadger = honeyBadgers.get(ipAddress)

      honeyBadgers.set(ipAddress, { ...honeyBadger, geoLocation: { ...rest } })
    })
  }

  /**
   * Creates a new stateful React component that is responsible for providing the "root" for the
   * visual and logical React components that make up the app.
   */
  constructor() {
    super()

    this.state = initialState
  }

  /**
   * React component lifecycle method that will be called after the <App> component has been mounted
   * into the DOM. After which we make calls to our GraphQL service/API endpoint to collect
   * information about honey badgers and update our state with it.
   */

  // It is okay to call 'setState()' in 'componentDidMount()' as discussed in the React docs
  /* eslint-disable react/no-did-mount-set-state */
  async componentDidMount() {
    try {
      // Query for the top honey badgers from our GraphQL service/API endpoint and enqueue a request
      // to set the apps state to the honey badgers
      const honeyBadgers = await App.collectTopHoneyBadgersInfo()

      this.setState({ honeyBadgers })

      // Now collect the following info about honey badgers from our GraphQL service/API endpoint
      // and enqueue a request to set the state of the app with the new info we collected:
      // * The autonomous systems that oversee the IP address of the honey badger
      // * The geospatial location of a honey badgers IP address
      await App.collectAutonomousSystemsInfo(honeyBadgers)
      this.setState({ honeyBadgers })

      await App.collectGeoLocationsInfo(honeyBadgers)
      this.setState({ honeyBadgers })
    } catch (error) {
      // It is okay to call 'setState()' in 'componentDidMount()' as discussed in the React docs
      // eslint-disable-next-line react/no-did-mount-set-state
      this.setState({ error })
    }
  }

  render() {
    const { honeyBadgers, error } = this.state

    if (error) {
      return `The following error occurred: ${error}`
    }

    return <HoneyBadgersTable honeyBadgers={[...honeyBadgers.values()]} />
  }
}

export default App
