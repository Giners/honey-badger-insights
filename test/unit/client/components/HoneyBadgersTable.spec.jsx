/* eslint prefer-arrow-callback: 0, func-names: 0 */
/* eslint no-unused-expressions: 0, chai-friendly/no-unused-expressions: 2 */

// General 3rd-party supporting test libs
import chai, { expect } from 'chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

// General 3rd-party supporting libs
import React from 'react'
import { Grid, Table } from '@devexpress/dx-react-grid-material-ui'

// App React component under test
import HoneyBadgersTable from './../../../../src/client/components/HoneyBadgersTable'

// Supporting app code
import { initialState } from './../../../../src/client/containers/App'

// Configure Chai to work with Jest
chai.use(chaiJestSnapshot)

describe('React component test: <HoneyBadgersTable>', function() {
  beforeEach(
    'Configure chai-jest-snapshot for "Mocha configuration mode"',
    function() {
      chaiJestSnapshot.configureUsingMochaContext(this)
      chaiJestSnapshot.setFilename(`${__filename}.snap`)
    },
  )

  describe('Renders correctly for given application state:', function() {
    // Reference to an Enzyme shallow wrapper around our <HoneyBadgersTable> component that can be
    // used throughout the following tests. Ought to be set before each test is ran.
    let hbtWrapper = null

    beforeEach('Setup Enzyme wrapper', function() {
      // Pass in props to the <HoneyBadgersTable> component the same the <App> component does when
      // it renders it.
      hbtWrapper = shallow(
        <HoneyBadgersTable
          honeyBadgers={[...initialState.honeyBadgers.values()]}
        />,
      )

      expect(hbtWrapper).to.not.be.null
      expect(hbtWrapper.exists()).to.be.true
    })

    it('Initial props/state', function() {
      // Even though the props/state gets set initially in our 'before' hooks we explicitly set it
      // here to be clear of the props/state that we expect to test
      hbtWrapper.setProps({
        honeyBadgers: [...initialState.honeyBadgers.values()],
      })

      // Ensure that we render a <Grid> and <Table> even if there is no data
      expect(hbtWrapper.find(Grid).exists()).to.be.true
      expect(hbtWrapper.find(Table).exists()).to.be.true

      // We shouldn't have any rows since we have no honey badgers. Lets double check...
      expect(hbtWrapper.find(Grid).prop('rows')).to.be.empty

      expect(toJson(hbtWrapper)).to.matchSnapshot()
    })

    it('Rows rendered (honey badgers data w/o additional data)', function() {
      // Create data as if we have only issued our initial GraphQL query of 'topHoneyBadgers'
      const honeyBadgers = [{ ipAddress: '1.1.1.1', count: '987' }]

      hbtWrapper.setProps({ honeyBadgers })

      // We should have some rows since we have some honey badgers. Check and ensure that each row
      // specifies it is loading data where appropriate
      const rows = hbtWrapper.find(Grid).prop('rows')
      expect(rows.length).to.equal(honeyBadgers.length)

      rows.forEach(row => {
        // This assertion is meant to remind us to update the following assertions. It will fail as
        // we add additional keys to the row data and don't update the tests. We currently expect
        // the following keys: 'ipAddress', 'count', 'asn'
        expect(Object.keys(row).length).to.equal(3)
        expect(row.ipAddress).to.not.be.empty
        expect(parseInt(row.count, 10)).to.be.above(0)

        // This test presumes we haven't yet queried for autonomous systems data so we expect it any
        // info we would display about autonomous systems to say 'Loading' at this point
        expect(row.asn).to.equal('Loading')
      })

      expect(toJson(hbtWrapper)).to.matchSnapshot()
    })

    it('Rows rendered (honey badgers data w/ autonomous systems data)', function() {
      // Create data as if we have issued the 'topHoneyBadgers' and 'autonomousSystems' queries to
      // our GraphQL service/API endpoint
      const honeyBadgers = [
        {
          ipAddress: '1.1.1.1',
          count: '987',
          as: {
            name: 'OVH SAS',
            asn: 16276,
            countryCode: 'FR',
          },
        },
      ]

      hbtWrapper.setProps({ honeyBadgers })

      // We should have some rows since we have some honey badgers. Check and ensure that each row
      // specifies it is loading data where appropriate
      const rows = hbtWrapper.find(Grid).prop('rows')
      expect(rows.length).to.equal(honeyBadgers.length)

      rows.forEach(row => {
        // This assertion is meant to remind us to update the following assertions. It will fail as
        // we add additional keys to the row data and don't update the tests. We currently expect
        // the following keys: 'ipAddress', 'count', 'asn'
        expect(Object.keys(row).length).to.equal(3)
        expect(row.ipAddress).to.not.be.empty
        expect(parseInt(row.count, 10)).to.be.above(0)
        expect(parseInt(row.asn, 10)).to.be.above(0)
      })

      expect(toJson(hbtWrapper)).to.matchSnapshot()
    })
  })
})
