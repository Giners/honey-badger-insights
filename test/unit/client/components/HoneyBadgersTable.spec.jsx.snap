// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React component test: <HoneyBadgersTable> Renders correctly for given application state: Initial props/state 1`] = `
<Grid$$1
  columns={
    Array [
      Object {
        "name": "ipAddress",
        "title": "IP Address",
      },
      Object {
        "name": "count",
        "title": "Times Seen Last 24 Hours",
      },
      Object {
        "name": "asn",
        "title": "Autonomous System #",
      },
    ]
  }
  rows={Array []}
>
  <PagingState
    defaultCurrentPage={0}
    defaultPageSize={10}
    pageSize={5}
  />
  <IntegratedPaging />
  <Table$$1
    messages={Object {}}
  />
  <TableHeaderRow$$1
    messages={Object {}}
  />
  <PagingPanel$$1
    messages={Object {}}
  />
</Grid$$1>
`;

exports[`React component test: <HoneyBadgersTable> Renders correctly for given application state: Rows rendered (honey badgers data w/ autonomous systems data) 1`] = `
<Grid$$1
  columns={
    Array [
      Object {
        "name": "ipAddress",
        "title": "IP Address",
      },
      Object {
        "name": "count",
        "title": "Times Seen Last 24 Hours",
      },
      Object {
        "name": "asn",
        "title": "Autonomous System #",
      },
    ]
  }
  rows={
    Array [
      Object {
        "asn": 16276,
        "count": "987",
        "ipAddress": "1.1.1.1",
      },
    ]
  }
>
  <PagingState
    defaultCurrentPage={0}
    defaultPageSize={10}
    pageSize={5}
  />
  <IntegratedPaging />
  <Table$$1
    messages={Object {}}
  />
  <TableHeaderRow$$1
    messages={Object {}}
  />
  <PagingPanel$$1
    messages={Object {}}
  />
</Grid$$1>
`;

exports[`React component test: <HoneyBadgersTable> Renders correctly for given application state: Rows rendered (honey badgers data w/o additional data) 1`] = `
<Grid$$1
  columns={
    Array [
      Object {
        "name": "ipAddress",
        "title": "IP Address",
      },
      Object {
        "name": "count",
        "title": "Times Seen Last 24 Hours",
      },
      Object {
        "name": "asn",
        "title": "Autonomous System #",
      },
    ]
  }
  rows={
    Array [
      Object {
        "asn": "Loading",
        "count": "987",
        "ipAddress": "1.1.1.1",
      },
    ]
  }
>
  <PagingState
    defaultCurrentPage={0}
    defaultPageSize={10}
    pageSize={5}
  />
  <IntegratedPaging />
  <Table$$1
    messages={Object {}}
  />
  <TableHeaderRow$$1
    messages={Object {}}
  />
  <PagingPanel$$1
    messages={Object {}}
  />
</Grid$$1>
`;
